File: Training/1.5 /1.5.1-ariprog.cpp
/*
ID:   andy1li
LANG: C++14
TASK: ariprog 
*/

#include <bits/stdc++.h>
using namespace std;

#define rep(i, j, k) for (size_t i=j; i<=k; ++i)
#define all(c) begin(c), end(c)
#define len(c) c.size()
#define each(x, c) for (auto& x: c)
#define pb push_back

/*----------------------------------------------------------------------------*/

int N, M;
bool bsq[2 * 251*251];

bool check(int a, int b) {
    rep(i, 0, N-1) {
        if (!bsq[a + b*i]) return false;
    }
    return true;
}

void solve() {
    cin >> N >> M;
    const int stop = 2 * M*M;

    rep(p, 0, M) rep(q, 0, M) {
        bsq[p*p + q*q] = true;
    }

    vector<pair<int, int>> ans;
    rep(a, 0, stop) {
        if (!bsq[a]) continue;
        rep(b, 1, (stop-a)/(N-1)) {
            if (check(a, b)) ans.pb({b, a});
        }
    }
    sort(all(ans));

    if (len(ans)) {
        each(pair, ans) cout << pair.second << ' ' << pair.first << endl;
    } else {
        cout << "NONE" << endl;
    } 
}

/*----------------------------------------------------------------------------*/

int main() {
    freopen("ariprog.in",  "r", stdin);
    freopen("ariprog.out", "w", stdout);
    solve();
}

File: Training/1.5 /1.5.1-ariprog.py
'''
ID:   andy1li
LANG: PYTHON3
TASK: ariprog
'''

from itertools import product

def solve():
    N, M = int(input()), int(input())
    sq = (i*i for i in range(M+1))
    check = frozenset(a + b for a, b in product(sq, repeat=2))
    bsq = sorted(check)
 
    ans = []
    for i in reversed(range(len(bsq))):
        for j in reversed(range(i)):
            last, pen = bsq[i], bsq[j]
            step = last - pen
            first = last - step * (N-1)
            if first < 0: break
            if all(first + step*i in check for i in range(N-2)):
                ans.append((first, step))
                
    ans = sorted(ans, key=lambda pair: (pair[1], pair[0]))

    print_line(
        '\n'.join(f'{pair[0]} {pair[1]}' for pair in ans) 
        if ans else 
        'NONE'
    )

#------------------------------------------------------------------------------#

def print_line(line): print(line); f_out.write( f'{line}\n' )

task = 'ariprog'
with open(task+'.in') as f_in, open(task+'.out', 'w') as f_out:
    input = f_in.readline
    solve()